highlight/HighlightsFromPointsExplainer.md

```diff
@@ -11,61 +11,549 @@ This document is intended as a starting point for engaging the community and sta
+        listenForMouseMove = false;
+    });
+</script>
+</body>
+</html>
+```
+
+### Performance Analysis
+
+We compared the performance of the two approaches presented for the online collaboration example analyzed throughout this explainer: with and without the proposed `highlightsFromPoint` API.
+We compared how long it takes for the event listener added for 'mousemove' events to complete in both cases when the mouse moves over a highlighted portion and when it moves over unaffected text. For this, we slightly modified the listener as shown below and also implemented a `testPerformance` function as follows, adding a hundred extra highlights to stress the test:
+
+```html
+let accumulatedTime = 0;
+let eventsFired = 0;
+div.addEventListener('mousemove', (event) => {
+    let start = performance.now();
+    if (listenForMouseMove) {
+        createActiveHighlights(event.pageX, event.pageY);
+    }
+    accumulatedTime += performance.now() - start;
+    eventsFired++;
+});
+
+function wait(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+async function testPerformance() {
+    let mousemoveEvent = new Event('mousemove');
+    let highlightRange = new Range();
+    highlightRange.setStart(bobRange.startContainer, bobRange.startOffset);
+    highlightRange.setEnd(bobRange.endContainer, bobRange.endOffset);
+    let rangeRect = highlightRange.getClientRects()[0];
+
+    mousemoveEvent.pageX = rangeRect.left + rangeRect.width/2;
+    mousemoveEvent.pageY = rangeRect.top + rangeRect.height/2;
+
+    // Add more highlights to stress the test.
+    for(let i=0; i<100; i++){
+        let h = new Highlight(bobRange);
+        let name = "highlight-" + i;
+        CSS.highlights.set(name, h);
+        highlightToUsername.set(h, name)
+    }
+    
+    // First dispatches can be outliers.
+    for(let i=0; i<100; i++){
+        div.dispatchEvent(mousemoveEvent);
+    }
+    await wait(100); 
+    
+    // Measure the time it takes to process a mousemove event over highlights.
+    accumulatedTime = 0;
+    eventsFired = 0;
+    for(let i=0; i<1000; i++){
+        div.dispatchEvent(mousemoveEvent);
+    }
+    await wait(1000); 
+    console.log("Average time it took to process a mousemove event over highlights: " + (accumulatedTime/eventsFired));
+
+    // Now measure the time it takes to process a mousemove event outside of the highlights.
+    mousemoveEvent.pageY = rangeRect.top - 1;
+    accumulatedTime = 0;
+    eventsFired = 0;
+    for(let i=0; i<1000; i++){
+        div.dispatchEvent(mousemoveEvent);
+    }
+    await wait(1000);
+    console.log("Average time it took to process a mousemove event outside of the highlights: " + (accumulatedTime/eventsFired));
+}
+
+testPerformance();
+```
+
+After executing both examples, we got the following results showing a ~1.9x improvement on hovering over highlights and a ~17x improvement on hovering outside of any highlights when using `highlightsFromPoint`:

```

Comment:
Consider calling out the results of the perf investigation somewhere higher up in the explainer, that way it is front and center for readers.

Do you have a sense of why using highlightsFromPoint is so much better? JS overhead? Garbage collection? Something else?
Offering an explanation on why we get perf wins can be important to convince readers of the validity of your experiment.

Reply:
I haven't had time to look at this perf section in detail yet -- will get to it as soon as I can -- when thinking about the co-editing scenario presented elsewhere in the explainer it seems like the fact that we have to use `mousemove` with `highlightsFromPoint` could be overall much worse for perf overall because we have to recalculate all the highlights **every time** the user moves the mouse in the entire editing area. But with the event-based model, those events should only fire, once, when the mouse enters or leaves some highlight.
So I don't think that example paints the `highlightsFromPoint` API in the best light.

For a scenario that reacts to clicking on the highlights, it could still be better though, since you won't need a performance-heavy mousemove event.

Reply:
+1 to Sanket's questions too :)